Fix extra Target configurations are generated when Project has custom configurations

francuim-d opened the pull request:
Resolves https://github.com/tuist/tuist/issues/4804

### Short description 📝

This Pull Request fixes the problem of generating extra configurations for the target. This case occurs when the `Target` does not have `settings` specified in the manifest, the `Project` has custom `settings` with custom configurations and there are module maps to be linked to a given target.

### How to test the changes locally 🧐

All existing tests pass. This case can be tested on an [example project](https://github.com/francuim-d/TuistConfigurationsSample). Here custom configurations are specified only for Project. When `.external(name: "FirebaseCrashlytics")` dependency added to Target extra settings are generated.

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] The title of the PR will be used as changelog entry, please make sure it is clear and suitable
- [x] In case the PR introduces changes that affect users, the documentation has been updated
- [x] Contributors have checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`

francuim-d commented:
Since the targets inherit the project settings, perhaps the best solution here is not to duplicate the project settings, but to create an empty dictionary, which will be updated with the necessary settings.

danyf90 commented:
Thanks @francuim-d ! 
can you add a unit test, and possible also enhance one of the available [fixtures](https://github.com/tuist/tuist/tree/main/projects/tuist/fixtures) to cover this scenario? 🙏

francuim-d commented:
@danyf90 I will try to add tests. 
What do you think about this? Should we leave the current implementation or rewrite to create empty settings?
> Since the targets inherit the project settings, perhaps the best solution here is not to duplicate the project settings, but to create an empty dictionary, which will be updated with the necessary settings.

danyf90 commented:
> What do you think about this? Should we leave the current implementation or rewrite to create empty settings?
> 

Yes, makes perfect sense 👌  Just ccing @kwridan for a second though as he's the master of Xcode settings :D

francuim-d commented:
I've rewrote code to use empty settings as a default and added a test case. But I have no idea which fixture and how it can be improved specifically for this case.

kwridan reviewed (changes requested):
Thanks for contributing this @francuim-d !

francuim-d commented on Sources/TuistGenerator/Mappers/ModuleMapMapper.swift:
```diff
@@ -101,7 +101,12 @@ public final class ModuleMapMapper: WorkspaceMapping {
+                    base: [:],
+                    configurations: [:],
+                    defaultSettings: .none
+                )
```
Thank you for review! I pushed changes related to this.

danyf90 commented on Sources/TuistGenerator/Mappers/ModuleMapMapper.swift:
```diff
@@ -101,7 +101,14 @@ public final class ModuleMapMapper: WorkspaceMapping {
                 }
 
-                mappedTarget.settings = (mappedTarget.settings ?? .default).with(base: mappedSettingsDictionary)
+                let defaultSettings = mappedProject.settings.defaultSettings == .none ? .recommended :
+                mappedProject.settings.defaultSettings
+                let targetSettings = mappedTarget.settings ?? Settings(
+                    base: [:],
+                    configurations: [:],
+                    defaultSettings: defaultSettings
+                )
```
Either I'm missing something, or this can be
```suggestion
                let targetSettings = mappedTarget.settings ?? Settings(
                    base: [:],
                    configurations: [:],
                    defaultSettings: mappedProject.settings.defaultSettings ?? .recommended
                )
```

francuim-d replied:
`.none` he is not an `Optional` case. So this syntax will not work here

kwridan replied:
ah good point - I mistakenly thought it was optional, in that case we can fallback to the project's default settings, no need to further fallback to recommend if it's `none` to maintain the existing behaviour.

Thanks!

francuim-d replied:
Fixed!

kwridan reviewed (approved):
Thanks @francuim-d
